src/cache/assignment-cache.ts

```diff
@@ -0,0 +1,81 @@
+import { getMD5Hash } from '../obfuscation';
+
+import { LRUCache } from './lru-cache';
+
+export type AssignmentCacheKey = {
+  subjectKey: string;
+  flagKey: string;
+  allocationKey: string;
+  variationKey: string;
+};
+
+export interface AsyncMap<K, V> {
+  get(key: K): Promise<V | undefined>;
+  set(key: K, value: V): Promise<void>;
+  has(key: K): Promise<boolean>;
+}
+
+export interface AssignmentCache {
+  set(key: AssignmentCacheKey): void;
+  has(key: AssignmentCacheKey): boolean;
+}
+
+export abstract class AssignmentCacheImpl<T extends Map<string, string>>
+  implements AssignmentCache
+{
+  // key -> variation value hash
+  protected constructor(protected readonly delegate: T) {}
+
+  has(key: AssignmentCacheKey): boolean {
+    const isPresent = this.delegate.has(this.toCacheKeyString(key));
+    if (!isPresent) {
+      // no cache key present
+      return false;
+    }
+
+    // the subject has been assigned to a different variation
+    // than was previously logged.
+    // in this case we need to log the assignment again.
+    const cachedValue = this.get(key);
+    return cachedValue === getMD5Hash(key.variationKey);
```

Comment:
I think we don't need the md5 anymore - when this was first implemented we were using the variation **_value_** which could be json of arbitrary length. I wanted to use the md5 to keep it a consistent length. However I see that along the way it was changed to the variation key.

ah ha.. by @schmit 😄 https://github.com/Eppo-exp/js-client-sdk-common/pull/45/files#diff-9154f1d3f50f8d666f2e17722cd162225cdd37359b799165dce470dd7d7c6ef8

Reply:
Revisiting this implementation - a `Map` might be over-kill and we are looking for a `Set` with the `variationKey` included. However it is fine as-is.

Reply:
Could users inspect (any of) the cache(s) from the client-side SDK? If so then perhaps we should leave it hashed

Reply:
either way is fine by me (map or set), both should be roughly equivalent

Reply:
@leoromanovsky should we remove `getMD5Hash` here?

Reply:
discussed offline, fine to keep as is